syweb/webclient/app/components/matrix/matrix-service.js

```diff
@@ -243,118 +273,161 @@ function($http, $window, $timeout, $q) {
-                    var firstRegType = undefined;
-                    
-                    for (var i=0; i<flows.length; i++) {
-                        var isThreePidFlow = false;
-                        if (flows[i].stages) {
-                            for (var j=0; j<flows[i].stages.length; j++) {
-                                var regType = flows[i].stages[j];
-                                if (knownTypes.indexOf(regType) === -1) {
-                                    deferred.reject("Unknown type: "+regType);
-                                    return;
-                                }
-                                if (regType == "m.login.email.identity") {
-                                    isThreePidFlow = true;
-                                }
-                                if (!useCaptcha && regType == "m.login.recaptcha") {
-                                    console.error("Web client setup to not use captcha, but HS demands a captcha.");
-                                    deferred.reject({
-                                        data: {
-                                            errcode: "M_CAPTCHA_NEEDED",
-                                            error: "Home server requires a captcha."
-                                        }
-                                    });
-                                    return;
-                                }
-                            }
-                        }
-                        
-                        if ( (isThreePidFlow && useThreePidFlow) || (!isThreePidFlow && !useThreePidFlow) ) {
-                            flowIndex = i;
-                        }
-                        
-                        if (knownTypes.indexOf(flows[i].type) == -1) {
-                            deferred.reject("Unknown type: "+flows[i].type);
-                            return;
+            if (bind_email === undefined) bind_email = false;
+
+            var data = {
+                username: user_name,
+                password: password,
+                bind_email: bind_email
+            };
+            if (sessionId) {
+                data['auth'] = {};
+                data['auth']['session'] = sessionId;
+            }
+            doV2Request("POST", path, undefined, data).then(function(resp) {
+                console.log("Got 2xx response straight away: completed!");
+                deferred.resolve(resp);
+            }, function(error) {
+                console.log("/register [1] : "+JSON.stringify(error));
+                if (error.data == undefined) {
+                    deferred.reject(error);
+                    return;
+                }
+                var flows = error.data.flows;
+                var params = error.data.params;
+                sessionId = error.data.session;
+                var completed = error.data.completed;
+
+                if (completed == undefined) {
+                    completed = [];
```

Comment:
This can be more succinctly represented as:
```javascript
var completed = error.data.completed || [];
```
which will set `completed` to `[]` if `error.data.completed` is falsey (`null`, `undefined`).